src/Bicep.Core/Syntax/ResourceDeclarationSyntax.cs

```diff
@@ -81,13 +81,32 @@ public TypeSymbol GetDeclaredType(IBinder binder, IResourceTypeProvider resource
             ResourceTypeReference? typeReference;
             var ancestors = binder.GetAllAncestors<ResourceDeclarationSyntax>(this);
             if (ancestors.Length == 0)
-            {
+            {                
                 // This is a top level resource - the type is a fully-qualified type.
                 typeReference = ResourceTypeReference.TryParse(stringContent);
                 if (typeReference == null)
                 {
                     return ErrorType.Create(DiagnosticBuilder.ForPosition(this.Type).InvalidResourceType());
                 }
+
+                if (binder.GetSymbolInfo(this) is ResourceSymbol resourceSymbol &&
+                    resourceSymbol.SafeGetBodyPropertyValue(LanguageConstants.ResourceParentPropertyName) is {} referenceParentSyntax &&
+                    binder.GetSymbolInfo(referenceParentSyntax) is ResourceSymbol parentResourceSymbol)
```

Comment:
I think we should block self-reference here. The following code causes stack overflow:
```
resource vmExt 'Microsoft.Compute/virtualMachines/extensions@2020-06-01' = {
  parent: vmExt // stack overflow
  location: 'eastus'
}
```

Reply:
And completions shouldn't offer `vmExt` as an accessible symbol. (It should happen without any changes, but I'd double check.)

Reply:
Oof, good catch. I expected the cycle detection logic to take care of that.